{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import tempfile\n",
    "import urllib.request\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyranges as pr\n",
    "\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_genes_and_transcripts(release=99, \n",
    "                                species='homo_sapiens',\n",
    "                                build='GRCh38',\n",
    "                                pub_build='',\n",
    "                                release2=None,\n",
    "                                datadir=Path('pyannotables') / Path('data'),\n",
    "                                join_gdf=None):\n",
    "\n",
    "    if release2 is None:\n",
    "        release2 = release\n",
    "\n",
    "    gtf_url = f'http://ftp.ensembl.org/pub/{pub_build.lower()}/release-{release}/gtf/{species}/{species.capitalize()}.{build}.{release2}.chr_patch_hapl_scaff.gtf.gz'\n",
    "\n",
    "    with tempfile.TemporaryDirectory() as tmpdir:\n",
    "        full_path = Path(tmpdir) / 'file.gtf.gz'\n",
    "        urllib.request.urlretrieve(gtf_url, full_path)\n",
    "        df = pr.read_gtf(full_path, output_df=True)\n",
    "    \n",
    "    gdf = df[df.Feature == 'gene'][['Chromosome', 'Source', 'Start', 'End', 'Strand',\n",
    "                                    'gene_id', 'gene_name', 'gene_source',\n",
    "                                    'gene_biotype']].drop_duplicates().set_index('gene_id')\n",
    "    if join_gdf is not None:\n",
    "        gdf = gdf.join(join_gdf)\n",
    "\n",
    "    tdf = df[df.Feature == 'transcript'][['Chromosome', 'Start', 'End', 'Strand',\n",
    "                                          'gene_id', 'transcript_id']].drop_duplicates().set_index('transcript_id')\n",
    "    datadir.mkdir(exist_ok=True)\n",
    "    gdf.to_pickle(datadir / f'datafile_{species}-{build}-ensembl{release}.pkl.xz')\n",
    "    tdf.to_pickle(datadir / f'datafile_{species}-{build}-ensembl{release}-tx2gene.pkl.xz')\n",
    "    \n",
    "    return gdf, tdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download HGNC gene name table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Alias symbols</th>\n",
       "      <th>Chromosome_region</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19q13.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268895</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NCRNA00181, A1BGAS, A1BG-AS</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>19q13.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000148584</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF, ASP, ACF64, ACF65, APOBEC1CF</td>\n",
       "      <td>10q11.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>12p13.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000245105</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12p13.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000162378</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>Approved</td>\n",
       "      <td>ZYG11</td>\n",
       "      <td>FLJ13456</td>\n",
       "      <td>1p32.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000159840</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7q34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274572</th>\n",
       "      <td>ZYXP1</td>\n",
       "      <td>zyxin pseudogene 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8q24.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000074755</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KIAA0399, ZZZ4, FLJ10821</td>\n",
       "      <td>17p13.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000036549</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DKFZP564I052, ATAC1</td>\n",
       "      <td>1p31.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36941 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Approved symbol  \\\n",
       "Ensembl gene ID                   \n",
       "ENSG00000121410            A1BG   \n",
       "ENSG00000268895        A1BG-AS1   \n",
       "ENSG00000148584            A1CF   \n",
       "ENSG00000175899             A2M   \n",
       "ENSG00000245105         A2M-AS1   \n",
       "...                         ...   \n",
       "ENSG00000162378          ZYG11B   \n",
       "ENSG00000159840             ZYX   \n",
       "ENSG00000274572           ZYXP1   \n",
       "ENSG00000074755           ZZEF1   \n",
       "ENSG00000036549            ZZZ3   \n",
       "\n",
       "                                                     Approved name    Status  \\\n",
       "Ensembl gene ID                                                                \n",
       "ENSG00000121410                             alpha-1-B glycoprotein  Approved   \n",
       "ENSG00000268895                               A1BG antisense RNA 1  Approved   \n",
       "ENSG00000148584                     APOBEC1 complementation factor  Approved   \n",
       "ENSG00000175899                              alpha-2-macroglobulin  Approved   \n",
       "ENSG00000245105                                A2M antisense RNA 1  Approved   \n",
       "...                                                            ...       ...   \n",
       "ENSG00000162378       zyg-11 family member B, cell cycle regulator  Approved   \n",
       "ENSG00000159840                                              zyxin  Approved   \n",
       "ENSG00000274572                                 zyxin pseudogene 1  Approved   \n",
       "ENSG00000074755  zinc finger ZZ-type and EF-hand domain contain...  Approved   \n",
       "ENSG00000036549                   zinc finger ZZ-type containing 3  Approved   \n",
       "\n",
       "                            Previous symbols  \\\n",
       "Ensembl gene ID                                \n",
       "ENSG00000121410                          NaN   \n",
       "ENSG00000268895  NCRNA00181, A1BGAS, A1BG-AS   \n",
       "ENSG00000148584                          NaN   \n",
       "ENSG00000175899                          NaN   \n",
       "ENSG00000245105                          NaN   \n",
       "...                                      ...   \n",
       "ENSG00000162378                        ZYG11   \n",
       "ENSG00000159840                          NaN   \n",
       "ENSG00000274572                          NaN   \n",
       "ENSG00000074755                          NaN   \n",
       "ENSG00000036549                          NaN   \n",
       "\n",
       "                                     Alias symbols Chromosome_region  \n",
       "Ensembl gene ID                                                       \n",
       "ENSG00000121410                                NaN          19q13.43  \n",
       "ENSG00000268895                           FLJ23569          19q13.43  \n",
       "ENSG00000148584  ACF, ASP, ACF64, ACF65, APOBEC1CF          10q11.23  \n",
       "ENSG00000175899             FWP007, S863-7, CPAMD5          12p13.31  \n",
       "ENSG00000245105                                NaN          12p13.31  \n",
       "...                                            ...               ...  \n",
       "ENSG00000162378                           FLJ13456            1p32.3  \n",
       "ENSG00000159840                                NaN              7q34  \n",
       "ENSG00000274572                                NaN           8q24.23  \n",
       "ENSG00000074755           KIAA0399, ZZZ4, FLJ10821           17p13.2  \n",
       "ENSG00000036549                DKFZP564I052, ATAC1            1p31.1  \n",
       "\n",
       "[36941 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = 'https://www.genenames.org/cgi-bin/download/custom?col=gd_app_sym&col=gd_app_name&col=gd_status&col=gd_prev_sym&col=gd_aliases&col=gd_pub_chrom_map&col=gd_pub_ensembl_id&status=Approved&status=Entry%20Withdrawn&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&submit=submit'\n",
    "human_genes = pd.read_table(url)\n",
    "human_genes = human_genes[~human_genes['Ensembl gene ID'].isnull()].set_index('Ensembl gene ID')\n",
    "human_genes.rename(columns={'Chromosome': 'Chromosome_region'}, inplace=True)\n",
    "human_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRCh38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8d83f077716b44c5bbad7bd87d7609f2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=3), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Alias symbols</th>\n",
       "      <th>Chromosome_region</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>havana</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASP family homolog 7, pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAM39F</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-34P13.3</td>\n",
       "      <td>havana</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2 host gene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274890</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome   Source  Start    End Strand     gene_name  \\\n",
       "gene_id                                                                  \n",
       "ENSG00000223972          1   havana  11869  14409      +       DDX11L1   \n",
       "ENSG00000227232          1   havana  14404  29570      -        WASH7P   \n",
       "ENSG00000278267          1  ensembl  17369  17436      -     MIR6859-1   \n",
       "ENSG00000243485          1   havana  29554  31109      +  RP11-34P13.3   \n",
       "ENSG00000274890          1  ensembl  30366  30503      +     MIR1302-2   \n",
       "\n",
       "                gene_source                        gene_biotype  \\\n",
       "gene_id                                                           \n",
       "ENSG00000223972      havana  transcribed_unprocessed_pseudogene   \n",
       "ENSG00000227232      havana              unprocessed_pseudogene   \n",
       "ENSG00000278267     ensembl                               miRNA   \n",
       "ENSG00000243485      havana                             lincRNA   \n",
       "ENSG00000274890     ensembl                               miRNA   \n",
       "\n",
       "                Approved symbol                      Approved name    Status  \\\n",
       "gene_id                                                                        \n",
       "ENSG00000223972         DDX11L1      DEAD/H-box helicase 11 like 1  Approved   \n",
       "ENSG00000227232          WASH7P  WASP family homolog 7, pseudogene  Approved   \n",
       "ENSG00000278267             NaN                                NaN       NaN   \n",
       "ENSG00000243485     MIR1302-2HG                MIR1302-2 host gene  Approved   \n",
       "ENSG00000274890             NaN                                NaN       NaN   \n",
       "\n",
       "                Previous symbols Alias symbols Chromosome_region  \n",
       "gene_id                                                           \n",
       "ENSG00000223972              NaN           NaN           1p36.33  \n",
       "ENSG00000227232              NaN        FAM39F           1p36.33  \n",
       "ENSG00000278267              NaN           NaN               NaN  \n",
       "ENSG00000243485              NaN           NaN           1p36.33  \n",
       "ENSG00000274890              NaN           NaN               NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328</th>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305</th>\n",
       "      <td>1</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000488147</th>\n",
       "      <td>1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000619216</th>\n",
       "      <td>1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000473358</th>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome  Start    End Strand          gene_id\n",
       "transcript_id                                                   \n",
       "ENST00000456328          1  11869  14409      +  ENSG00000223972\n",
       "ENST00000450305          1  12010  13670      +  ENSG00000223972\n",
       "ENST00000488147          1  14404  29570      -  ENSG00000227232\n",
       "ENST00000619216          1  17369  17436      -  ENSG00000278267\n",
       "ENST00000473358          1  29554  31097      +  ENSG00000243485"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Alias symbols</th>\n",
       "      <th>Chromosome_region</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>havana</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASP family homolog 7, pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAM39F</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>1</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>havana</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2 host gene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>1</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>microRNA 1302-2</td>\n",
       "      <td>Approved</td>\n",
       "      <td>MIRN1302-2</td>\n",
       "      <td>hsa-mir-1302-2</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome   Source  Start    End Strand    gene_name  \\\n",
       "gene_id                                                                 \n",
       "ENSG00000223972          1   havana  11869  14409      +      DDX11L1   \n",
       "ENSG00000227232          1   havana  14404  29570      -       WASH7P   \n",
       "ENSG00000278267          1  mirbase  17369  17436      -    MIR6859-1   \n",
       "ENSG00000243485          1   havana  29554  31109      +  MIR1302-2HG   \n",
       "ENSG00000284332          1  mirbase  30366  30503      +    MIR1302-2   \n",
       "\n",
       "                gene_source                        gene_biotype  \\\n",
       "gene_id                                                           \n",
       "ENSG00000223972      havana  transcribed_unprocessed_pseudogene   \n",
       "ENSG00000227232      havana              unprocessed_pseudogene   \n",
       "ENSG00000278267     mirbase                               miRNA   \n",
       "ENSG00000243485      havana                             lincRNA   \n",
       "ENSG00000284332     mirbase                               miRNA   \n",
       "\n",
       "                Approved symbol                      Approved name    Status  \\\n",
       "gene_id                                                                        \n",
       "ENSG00000223972         DDX11L1      DEAD/H-box helicase 11 like 1  Approved   \n",
       "ENSG00000227232          WASH7P  WASP family homolog 7, pseudogene  Approved   \n",
       "ENSG00000278267             NaN                                NaN       NaN   \n",
       "ENSG00000243485     MIR1302-2HG                MIR1302-2 host gene  Approved   \n",
       "ENSG00000284332       MIR1302-2                    microRNA 1302-2  Approved   \n",
       "\n",
       "                Previous symbols   Alias symbols Chromosome_region  \n",
       "gene_id                                                             \n",
       "ENSG00000223972              NaN             NaN           1p36.33  \n",
       "ENSG00000227232              NaN          FAM39F           1p36.33  \n",
       "ENSG00000278267              NaN             NaN               NaN  \n",
       "ENSG00000243485              NaN             NaN           1p36.33  \n",
       "ENSG00000284332       MIRN1302-2  hsa-mir-1302-2           1p36.33  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328</th>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305</th>\n",
       "      <td>1</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000488147</th>\n",
       "      <td>1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000619216</th>\n",
       "      <td>1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000473358</th>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome  Start    End Strand          gene_id\n",
       "transcript_id                                                   \n",
       "ENST00000456328          1  11869  14409      +  ENSG00000223972\n",
       "ENST00000450305          1  12010  13670      +  ENSG00000223972\n",
       "ENST00000488147          1  14404  29570      -  ENSG00000227232\n",
       "ENST00000619216          1  17369  17436      -  ENSG00000278267\n",
       "ENST00000473358          1  29554  31097      +  ENSG00000243485"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Alias symbols</th>\n",
       "      <th>Chromosome_region</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>havana</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASP family homolog 7, pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAM39F</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>1</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>havana</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2 host gene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>1</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>mirbase</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>microRNA 1302-2</td>\n",
       "      <td>Approved</td>\n",
       "      <td>MIRN1302-2</td>\n",
       "      <td>hsa-mir-1302-2</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome   Source  Start    End Strand    gene_name  \\\n",
       "gene_id                                                                 \n",
       "ENSG00000223972          1   havana  11869  14409      +      DDX11L1   \n",
       "ENSG00000227232          1   havana  14404  29570      -       WASH7P   \n",
       "ENSG00000278267          1  mirbase  17369  17436      -    MIR6859-1   \n",
       "ENSG00000243485          1   havana  29554  31109      +  MIR1302-2HG   \n",
       "ENSG00000284332          1  mirbase  30366  30503      +    MIR1302-2   \n",
       "\n",
       "                gene_source                        gene_biotype  \\\n",
       "gene_id                                                           \n",
       "ENSG00000223972      havana  transcribed_unprocessed_pseudogene   \n",
       "ENSG00000227232      havana              unprocessed_pseudogene   \n",
       "ENSG00000278267     mirbase                               miRNA   \n",
       "ENSG00000243485      havana                              lncRNA   \n",
       "ENSG00000284332     mirbase                               miRNA   \n",
       "\n",
       "                Approved symbol                      Approved name    Status  \\\n",
       "gene_id                                                                        \n",
       "ENSG00000223972         DDX11L1      DEAD/H-box helicase 11 like 1  Approved   \n",
       "ENSG00000227232          WASH7P  WASP family homolog 7, pseudogene  Approved   \n",
       "ENSG00000278267             NaN                                NaN       NaN   \n",
       "ENSG00000243485     MIR1302-2HG                MIR1302-2 host gene  Approved   \n",
       "ENSG00000284332       MIR1302-2                    microRNA 1302-2  Approved   \n",
       "\n",
       "                Previous symbols   Alias symbols Chromosome_region  \n",
       "gene_id                                                             \n",
       "ENSG00000223972              NaN             NaN           1p36.33  \n",
       "ENSG00000227232              NaN          FAM39F           1p36.33  \n",
       "ENSG00000278267              NaN             NaN               NaN  \n",
       "ENSG00000243485              NaN             NaN           1p36.33  \n",
       "ENSG00000284332       MIRN1302-2  hsa-mir-1302-2           1p36.33  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328</th>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305</th>\n",
       "      <td>1</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000488147</th>\n",
       "      <td>1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000619216</th>\n",
       "      <td>1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000473358</th>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome  Start    End Strand          gene_id\n",
       "transcript_id                                                   \n",
       "ENST00000456328          1  11869  14409      +  ENSG00000223972\n",
       "ENST00000450305          1  12010  13670      +  ENSG00000223972\n",
       "ENST00000488147          1  14404  29570      -  ENSG00000227232\n",
       "ENST00000619216          1  17369  17436      -  ENSG00000278267\n",
       "ENST00000473358          1  29554  31097      +  ENSG00000243485"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "species = 'homo_sapiens'\n",
    "\n",
    "for release in tqdm((84, 93, 99)):\n",
    "    gdf, tdf = fetch_genes_and_transcripts(release, species, join_gdf=human_genes)\n",
    "    \n",
    "    display(gdf.head())\n",
    "    display(tdf.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRCh37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "release = 99\n",
    "release2 = 87\n",
    "\n",
    "gdf, tdf = fetch_genes_and_transcripts(release, species, build='GRCh37', pub_build='GRCh37', release2=release2, join_gdf=human_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Alias symbols</th>\n",
       "      <th>Chromosome_region</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASP family homolog 7, pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAM39F</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2 host gene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>-</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>family with sequence similarity 138 member A</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F379</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>+</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>olfactory receptor family 4 subfamily G member...</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1p36.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome          Source  Start    End Strand   gene_name  \\\n",
       "gene_id                                                                       \n",
       "ENSG00000223972          1  ensembl_havana  11869  14412      +     DDX11L1   \n",
       "ENSG00000227232          1  ensembl_havana  14363  29806      -      WASH7P   \n",
       "ENSG00000243485          1  ensembl_havana  29554  31109      +  MIR1302-10   \n",
       "ENSG00000237613          1  ensembl_havana  34554  36081      -     FAM138A   \n",
       "ENSG00000268020          1  ensembl_havana  52473  54936      +      OR4G4P   \n",
       "\n",
       "                    gene_source gene_biotype Approved symbol  \\\n",
       "gene_id                                                        \n",
       "ENSG00000223972  ensembl_havana   pseudogene         DDX11L1   \n",
       "ENSG00000227232  ensembl_havana   pseudogene          WASH7P   \n",
       "ENSG00000243485  ensembl_havana      lincRNA     MIR1302-2HG   \n",
       "ENSG00000237613  ensembl_havana      lincRNA         FAM138A   \n",
       "ENSG00000268020  ensembl_havana   pseudogene          OR4G4P   \n",
       "\n",
       "                                                     Approved name    Status  \\\n",
       "gene_id                                                                        \n",
       "ENSG00000223972                      DEAD/H-box helicase 11 like 1  Approved   \n",
       "ENSG00000227232                  WASP family homolog 7, pseudogene  Approved   \n",
       "ENSG00000243485                                MIR1302-2 host gene  Approved   \n",
       "ENSG00000237613       family with sequence similarity 138 member A  Approved   \n",
       "ENSG00000268020  olfactory receptor family 4 subfamily G member...  Approved   \n",
       "\n",
       "                Previous symbols Alias symbols Chromosome_region  \n",
       "gene_id                                                           \n",
       "ENSG00000223972              NaN           NaN           1p36.33  \n",
       "ENSG00000227232              NaN        FAM39F           1p36.33  \n",
       "ENSG00000243485              NaN           NaN           1p36.33  \n",
       "ENSG00000237613              NaN          F379           1p36.33  \n",
       "ENSG00000268020              NaN           NaN           1p36.33  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mouse annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRCm38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf, tdf = fetch_genes_and_transcripts(99, 'mus_musculus', build='GRCm38')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102693</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064842</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102851</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>3252757</td>\n",
       "      <td>3253236</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000103377</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>3365731</td>\n",
       "      <td>3368549</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm37180</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Chromosome          Source    Start      End Strand  \\\n",
       "gene_id                                                                  \n",
       "ENSMUSG00000102693          1          havana  3073253  3074322      +   \n",
       "ENSMUSG00000064842          1         ensembl  3102016  3102125      +   \n",
       "ENSMUSG00000051951          1  ensembl_havana  3205901  3671498      -   \n",
       "ENSMUSG00000102851          1          havana  3252757  3253236      +   \n",
       "ENSMUSG00000103377          1          havana  3365731  3368549      -   \n",
       "\n",
       "                        gene_name     gene_source          gene_biotype  \n",
       "gene_id                                                                  \n",
       "ENSMUSG00000102693  4933401J01Rik          havana                   TEC  \n",
       "ENSMUSG00000064842        Gm26206         ensembl                 snRNA  \n",
       "ENSMUSG00000051951           Xkr4  ensembl_havana        protein_coding  \n",
       "ENSMUSG00000102851        Gm18956          havana  processed_pseudogene  \n",
       "ENSMUSG00000103377        Gm37180          havana                   TEC  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000193812</th>\n",
       "      <td>1</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000082908</th>\n",
       "      <td>1</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000162897</th>\n",
       "      <td>1</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3216344</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000159265</th>\n",
       "      <td>1</td>\n",
       "      <td>3206523</td>\n",
       "      <td>3215632</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000070533</th>\n",
       "      <td>1</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3671498</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Chromosome    Start      End Strand             gene_id\n",
       "transcript_id                                                             \n",
       "ENSMUST00000193812          1  3073253  3074322      +  ENSMUSG00000102693\n",
       "ENSMUST00000082908          1  3102016  3102125      +  ENSMUSG00000064842\n",
       "ENSMUST00000162897          1  3205901  3216344      -  ENSMUSG00000051951\n",
       "ENSMUST00000159265          1  3206523  3215632      -  ENSMUSG00000051951\n",
       "ENSMUST00000070533          1  3214482  3671498      -  ENSMUSG00000051951"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdf.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
