{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from pyensembl import EnsemblRelease, genome_for_reference_name\n",
    "from pyensembl.species import human, mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_genes_and_transcripts(release, species, datadir=Path('pyannotables') / Path('data')):\n",
    "    er = EnsemblRelease(release=release, species=species)\n",
    "    er.download()\n",
    "    er.index()\n",
    "    \n",
    "    genes = er.genes()\n",
    "    tx = er.transcripts()\n",
    "\n",
    "    gdf = pd.DataFrame([x.to_dict() for x in genes])\n",
    "    gdf = gdf[['gene_id', 'gene_name', 'contig', 'start', 'end', 'strand', 'biotype']]\n",
    "    gdf.set_index('gene_id', inplace=True)\n",
    "\n",
    "    tdf = pd.DataFrame([x.to_dict() for x in tx])\n",
    "    tdf = tdf[['gene_id', 'transcript_id']]\n",
    "    tdf.set_index('transcript_id', inplace=True)\n",
    "\n",
    "    datadir.mkdir(exist_ok=True)\n",
    "    gdf.to_pickle(datadir / f'datafile_{species.latin_name}-ensembl{release}-{ref}.pkl.xz')\n",
    "    tdf.to_pickle(datadir / f'datafile_{species.latin_name}-ensembl{release}-{ref}-tx2gene.pkl.xz')\n",
    "    \n",
    "    return gdf, tdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download HGNC gene name table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NCRNA00181, A1BGAS, A1BG-AS</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF, ASP, ACF64, ACF65, APOBEC1CF</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1S9T</td>\n",
       "      <td>symbol withdrawn, see [HGNC:12469](/data/gene-...</td>\n",
       "      <td>Symbol Withdrawn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46814</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>Approved</td>\n",
       "      <td>ZYG11</td>\n",
       "      <td>FLJ13456</td>\n",
       "      <td>1p32.3</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46815</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7q34</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46816</th>\n",
       "      <td>ZYXP1</td>\n",
       "      <td>zyxin pseudogene 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8q24.23</td>\n",
       "      <td>ENSG00000274572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46817</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KIAA0399, ZZZ4, FLJ10821</td>\n",
       "      <td>17p13.2</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46818</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DKFZP564I052, ATAC1</td>\n",
       "      <td>1p31.1</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46819 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Approved symbol                                      Approved name  \\\n",
       "0                A1BG                             alpha-1-B glycoprotein   \n",
       "1            A1BG-AS1                               A1BG antisense RNA 1   \n",
       "2                A1CF                     APOBEC1 complementation factor   \n",
       "3               A1S9T  symbol withdrawn, see [HGNC:12469](/data/gene-...   \n",
       "4                 A2M                              alpha-2-macroglobulin   \n",
       "...               ...                                                ...   \n",
       "46814          ZYG11B       zyg-11 family member B, cell cycle regulator   \n",
       "46815             ZYX                                              zyxin   \n",
       "46816           ZYXP1                                 zyxin pseudogene 1   \n",
       "46817           ZZEF1  zinc finger ZZ-type and EF-hand domain contain...   \n",
       "46818            ZZZ3                   zinc finger ZZ-type containing 3   \n",
       "\n",
       "                 Status             Previous symbols  \\\n",
       "0              Approved                          NaN   \n",
       "1              Approved  NCRNA00181, A1BGAS, A1BG-AS   \n",
       "2              Approved                          NaN   \n",
       "3      Symbol Withdrawn                          NaN   \n",
       "4              Approved                          NaN   \n",
       "...                 ...                          ...   \n",
       "46814          Approved                        ZYG11   \n",
       "46815          Approved                          NaN   \n",
       "46816          Approved                          NaN   \n",
       "46817          Approved                          NaN   \n",
       "46818          Approved                          NaN   \n",
       "\n",
       "                                Synonyms Chromosome  Ensembl gene ID  \n",
       "0                                    NaN   19q13.43  ENSG00000121410  \n",
       "1                               FLJ23569   19q13.43  ENSG00000268895  \n",
       "2      ACF, ASP, ACF64, ACF65, APOBEC1CF   10q11.23  ENSG00000148584  \n",
       "3                                    NaN        NaN              NaN  \n",
       "4                 FWP007, S863-7, CPAMD5   12p13.31  ENSG00000175899  \n",
       "...                                  ...        ...              ...  \n",
       "46814                           FLJ13456     1p32.3  ENSG00000162378  \n",
       "46815                                NaN       7q34  ENSG00000159840  \n",
       "46816                                NaN    8q24.23  ENSG00000274572  \n",
       "46817           KIAA0399, ZZZ4, FLJ10821    17p13.2  ENSG00000074755  \n",
       "46818                DKFZP564I052, ATAC1     1p31.1  ENSG00000036549  \n",
       "\n",
       "[46819 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = 'https://www.genenames.org/cgi-bin/download/custom?col=gd_app_sym&col=gd_app_name&col=gd_status&col=gd_prev_sym&col=gd_aliases&col=gd_pub_chrom_map&col=gd_pub_ensembl_id&status=Approved&status=Entry%20Withdrawn&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&submit=submit'\n",
    "human_genes = pd.read_table(url)\n",
    "human_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRCh38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCh38/ensembl97/Homo_sapiens.GRCh38.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCh38/ensembl97/Homo_sapiens.GRCh38.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCh38/ensembl97/Homo_sapiens.GRCh38.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "release = 97\n",
    "species = human\n",
    "\n",
    "ref = species.which_reference(release)\n",
    "gdf, tdf = fetch_genes_and_transcripts(release, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>X</td>\n",
       "      <td>100627109</td>\n",
       "      <td>100639991</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>X</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>20</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50958555</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169894267</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name contig      start        end strand         biotype\n",
       "gene_id                                                                      \n",
       "ENSG00000000003    TSPAN6      X  100627109  100639991      -  protein_coding\n",
       "ENSG00000000005      TNMD      X  100584936  100599885      +  protein_coding\n",
       "ENSG00000000419      DPM1     20   50934867   50958555      -  protein_coding\n",
       "ENSG00000000457     SCYL3      1  169849631  169894267      -  protein_coding\n",
       "ENSG00000000460  C1orf112      1  169662007  169854080      +  protein_coding"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>biotype</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>X</td>\n",
       "      <td>100627109</td>\n",
       "      <td>100639991</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>tetraspanin 6</td>\n",
       "      <td>Approved</td>\n",
       "      <td>TM4SF6</td>\n",
       "      <td>T245, TSPAN-6</td>\n",
       "      <td>Xq22.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>X</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>tenomodulin</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>myodulin, ChM1L, tendin, TEM, BRICD4</td>\n",
       "      <td>Xq22.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>20</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50958555</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>dolichyl-phosphate mannosyltransferase subunit...</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MPDS, CDGIE</td>\n",
       "      <td>20q13.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169894267</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>SCY1 like pseudokinase 3</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PACE-1, PACE1</td>\n",
       "      <td>1q24.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>chromosome 1 open reading frame 112</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLJ10706</td>\n",
       "      <td>1q24.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name contig      start        end strand         biotype  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000000003    TSPAN6      X  100627109  100639991      -  protein_coding   \n",
       "ENSG00000000005      TNMD      X  100584936  100599885      +  protein_coding   \n",
       "ENSG00000000419      DPM1     20   50934867   50958555      -  protein_coding   \n",
       "ENSG00000000457     SCYL3      1  169849631  169894267      -  protein_coding   \n",
       "ENSG00000000460  C1orf112      1  169662007  169854080      +  protein_coding   \n",
       "\n",
       "                Approved symbol  \\\n",
       "gene_id                           \n",
       "ENSG00000000003          TSPAN6   \n",
       "ENSG00000000005            TNMD   \n",
       "ENSG00000000419            DPM1   \n",
       "ENSG00000000457           SCYL3   \n",
       "ENSG00000000460        C1orf112   \n",
       "\n",
       "                                                     Approved name    Status  \\\n",
       "gene_id                                                                        \n",
       "ENSG00000000003                                      tetraspanin 6  Approved   \n",
       "ENSG00000000005                                        tenomodulin  Approved   \n",
       "ENSG00000000419  dolichyl-phosphate mannosyltransferase subunit...  Approved   \n",
       "ENSG00000000457                           SCY1 like pseudokinase 3  Approved   \n",
       "ENSG00000000460                chromosome 1 open reading frame 112  Approved   \n",
       "\n",
       "                Previous symbols                              Synonyms  \\\n",
       "gene_id                                                                  \n",
       "ENSG00000000003           TM4SF6                         T245, TSPAN-6   \n",
       "ENSG00000000005              NaN  myodulin, ChM1L, tendin, TEM, BRICD4   \n",
       "ENSG00000000419              NaN                           MPDS, CDGIE   \n",
       "ENSG00000000457              NaN                         PACE-1, PACE1   \n",
       "ENSG00000000460              NaN                              FLJ10706   \n",
       "\n",
       "                Chromosome  \n",
       "gene_id                     \n",
       "ENSG00000000003     Xq22.1  \n",
       "ENSG00000000005     Xq22.1  \n",
       "ENSG00000000419   20q13.13  \n",
       "ENSG00000000457     1q24.2  \n",
       "ENSG00000000460     1q24.2  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf = gdf.merge(human_genes, how='left', left_index=True, right_on='Ensembl gene ID').set_index(gdf.index).drop('Ensembl gene ID', axis=1)\n",
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000000233</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000000412</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000000442</th>\n",
       "      <td>ENSG00000173153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000001008</th>\n",
       "      <td>ENSG00000004478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000001146</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gene_id\n",
       "transcript_id                   \n",
       "ENST00000000233  ENSG00000004059\n",
       "ENST00000000412  ENSG00000003056\n",
       "ENST00000000442  ENSG00000173153\n",
       "ENST00000001008  ENSG00000004478\n",
       "ENST00000001146  ENSG00000003137"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRCh37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "release = 75\n",
    "species = human\n",
    "\n",
    "ref = species.which_reference(release)\n",
    "gdf, tdf = fetch_genes_and_transcripts(release, species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mouse annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRCm38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCm38/ensembl97/Mus_musculus.GRCm38.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCm38/ensembl97/Mus_musculus.GRCm38.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/gokcen/Library/Caches/pyensembl/GRCm38/ensembl97/Mus_musculus.GRCm38.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "release = 97\n",
    "species = mouse\n",
    "\n",
    "ref = species.which_reference(release)\n",
    "gdf, tdf = fetch_genes_and_transcripts(release, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>108107280</td>\n",
       "      <td>108146146</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>77837901</td>\n",
       "      <td>77853623</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>18780447</td>\n",
       "      <td>18811987</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>H19</td>\n",
       "      <td>7</td>\n",
       "      <td>142575529</td>\n",
       "      <td>142578143</td>\n",
       "      <td>-</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>161082525</td>\n",
       "      <td>161258213</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_name contig      start        end strand  \\\n",
       "gene_id                                                            \n",
       "ENSMUSG00000000001     Gnai3      3  108107280  108146146      -   \n",
       "ENSMUSG00000000003      Pbsn      X   77837901   77853623      -   \n",
       "ENSMUSG00000000028     Cdc45     16   18780447   18811987      -   \n",
       "ENSMUSG00000000031       H19      7  142575529  142578143      -   \n",
       "ENSMUSG00000000037     Scml2      X  161082525  161258213      +   \n",
       "\n",
       "                           biotype  \n",
       "gene_id                             \n",
       "ENSMUSG00000000001  protein_coding  \n",
       "ENSMUSG00000000003  protein_coding  \n",
       "ENSMUSG00000000028  protein_coding  \n",
       "ENSMUSG00000000031          lncRNA  \n",
       "ENSMUSG00000000037  protein_coding  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000000001</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000000003</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000000010</th>\n",
       "      <td>ENSMUSG00000020875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000000028</th>\n",
       "      <td>ENSMUSG00000000028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000000033</th>\n",
       "      <td>ENSMUSG00000048583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               gene_id\n",
       "transcript_id                         \n",
       "ENSMUST00000000001  ENSMUSG00000000001\n",
       "ENSMUST00000000003  ENSMUSG00000000003\n",
       "ENSMUST00000000010  ENSMUSG00000020875\n",
       "ENSMUST00000000028  ENSMUSG00000000028\n",
       "ENSMUST00000000033  ENSMUSG00000048583"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdf.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
